client2/send.go

```diff
@@ -0,0 +1,162 @@
+// SPDX-FileCopyrightText: Copyright (C) 2017  Yawning Angel.
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package client2
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/katzenpost/katzenpost/core/crypto/rand"
+	"github.com/katzenpost/katzenpost/core/epochtime"
+	cpki "github.com/katzenpost/katzenpost/core/pki"
+	sConstants "github.com/katzenpost/katzenpost/core/sphinx/constants"
+	"github.com/katzenpost/katzenpost/core/sphinx/path"
+)
+
+// SendSphinxPacket sends the given Sphinx packet.
+func (c *Client) SendSphinxPacket(pkt []byte) error {
+	return c.conn.sendPacket(pkt)
+}
+
+// ComposeSphinxPacket is used to compose Sphinx packets.
+func (c *Client) ComposeSphinxPacket(request *Request) ([]byte, []byte, time.Duration, error) {
+
+	if len(request.RecipientQueueID) > sConstants.RecipientIDLength {
+		return nil, nil, 0, fmt.Errorf("client2: invalid recipient: '%v'", request.RecipientQueueID)
+	}
+
+	if len(request.Payload) > c.geo.UserForwardPayloadLength {
+		return nil, nil, 0, fmt.Errorf("message too large: %v > %v", len(request.Payload), c.geo.UserForwardPayloadLength)
+	}
+
+	payload := make([]byte, c.geo.UserForwardPayloadLength)
+	copy(payload, request.Payload)
+
+	for {
+		unixTime := c.pki.skewedUnixTime()
+		_, _, budget := epochtime.FromUnix(unixTime)
+		start := time.Now()
+
+		// Select the forward path.
+		now := time.Unix(unixTime, 0)
+
+		if c.conn.provider == nil {
+			panic("source provider cannot be nil")
+		}
+
+		fwdPath, then, err := c.makePath(request.RecipientQueueID, request.DestinationIdHash, request.SURBID, now, true)
+		if err != nil {
+			return nil, nil, 0, err
+		}
+
+		revPath := make([]*path.PathHop, 0)
+		if request.SURBID != nil {
```

Comment:
if the SURBID is not `nil`, won't it expire if we hit a wrap-around? (aren't SURBs tied to epoch?)
if so, that's another argument for relying on `epoch` as an argument to the function or property of `request` rather than `time.Now()` after a bunch of operations.